src/ops/noop.cc

```diff
@@ -166,23 +162,25 @@ Node FFModel::get_or_create_noop_node(const ParallelTensor input)
     cached_noop_ops[hash] = noop;
   }
   Node ret;
-  ret.guid = node_global_guid ++;
+  ret.guid = node_global_guid++;
   ret.ptr = noop;
   return ret;
 }
 
-Node FFModel::get_or_create_input_node(const ParallelTensorShape& output_shape)
-{
-  size_t hash = std::hash<ParallelTensorShape>{}(output_shape); 
-  NoOp* input = NULL;
-  const auto& it = cached_input_ops.find(hash);
+Node FFModel::get_or_create_input_node(
+    const ParallelTensorShape& output_shape) {
+  NoOp* input = nullptr;
+
+  // For OP_INPUT, output_shape is what distinguishes them.
+  const auto& it = cached_input_ops.find(output_shape);
   if (it != cached_input_ops.end()) {
     input = it->second;
   } else {
     ParallelTensor tensor = new ParallelTensorBase();
     tensor->parallel_tensor_guid = parallel_tensor_global_guid++;
-    tensor->data_type = DT_FLOAT; // TODO FIXME @lockshaw
+    tensor->data_type = DT_FLOAT;  // TODO FIXME @lockshaw
```

Comment:
While we're modifying `NoOp`, could you check if we can throw an actual assertion here on the datatype rather than just a TODO?